sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanData.java

```diff
@@ -0,0 +1,386 @@
+/*
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.trace.export;
+
+import com.google.auto.value.AutoValue;
+import io.opentelemetry.internal.Utils;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.trace.AttributeValue;
+import io.opentelemetry.trace.Span.Kind;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.SpanId;
+import io.opentelemetry.trace.Status;
+import io.opentelemetry.trace.Timestamp;
+import io.opentelemetry.trace.TraceFlags;
+import io.opentelemetry.trace.TraceId;
+import io.opentelemetry.trace.Tracestate;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Immutable representation of all data collected by the {@link io.opentelemetry.trace.Span} class.
+ *
+ * @since 0.1.0
+ */
+@Immutable
+@AutoValue
+public abstract class SpanData {
+
+  /**
+   * Returns a new immutable {@code SpanData}.
+   *
+   * @param context the {@code SpanContext} of the {@code Span}.
+   * @param parentSpanId the parent {@code SpanId} of the {@code Span}. {@code null} if the {@code
+   *     Span} is a root.
+   * @param resource the resource this span was executed on.
+   * @param name the name of the {@code Span}.
+   * @param kind the kind of the {@code Span}.
+   * @param startTimestamp the start {@code Timestamp} of the {@code Span}.
+   * @param attributes the attributes associated with the {@code Span}.
+   * @param timedEvents the events associated with the {@code Span}.
+   * @param links the links associated with the {@code Span}.
+   * @param status the {@code Status} of the {@code Span}.
+   * @param endTimestamp the end {@code Timestamp} of the {@code Span}.
+   * @return a new immutable {@code SpanData}.
+   * @since 0.1.0
+   */
+  public static SpanData create(
```

Comment:
Why create if we have a builder? just remove this.

Reply:
There's a little bit of helpful destructuring of the context and some defaulting/protecting of the attributes, links, and events collections.  Seems like there's enough value in there that we shouldn't let it go.

My instinct would be to put move that `create()` to be the `build()` method of the `Builder` but AutoValue makes that challenging doesn't it?  Thoughts?

Reply:
For the immutable copies see https://github.com/google/auto/blob/master/value/userguide/builders-howto.md#-normalize-modify-a-property-value-at-build-time. Use autoBuild() and write your own build.

For the SpanContext you can have an implemented setter on the Builder that calls setTraceId/set... Not sure if that is needed, in my opinion that can be done by the caller.

Reply:
cool. we'll make it so. I didn't know about that technique (I'm admittedly pretty new to AutoValue)